### PR TITLE
apps/system/utils/heapinfo : Support heapinfo for binary separation

### PR DESCRIPTION
**Pull request by jeongchanKim**
1. change getting user heap with temp allocated memory
2. show whole task/pthread current and peak heap usage
3. add initialization for total and peak alloc size of user heap

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [eab7ebe3ccc62c022d0597da7c2227dca12d0735] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [eab7ebe3ccc62c022d0597da7c2227dca12d0735] - Code Rule Check (C++) OK.

File: apps/system/utils/utils_heapinfo.c [hunk 1]
```diff
@@ -218,13 +216,7 @@ static void heapinfo_show_taskinfo(struct mm_heap_s *heap)
...
 	printf("-------|-----------|-----------|----------\n");
 
-	for (heap_idx = 0; heap_idx < CONFIG_MM_NHEAPS; heap_idx++) {
-		for (tcb_idx = 0; tcb_idx < CONFIG_MAX_TASKS; tcb_idx++) {
-			if (heap[heap_idx].alloc_list[tcb_idx].pid != HEAPINFO_INIT_INFO) {
-				heapinfo_read_proc(heap[heap_idx].alloc_list[tcb_idx].pid);
-			}
-		}
-	}
+	utils_proc_pid_foreach(heapinfo_read_proc);
```
Comment (Taejun-Kwon) [inline apps/system/utils/utils_heapinfo.c]: good. =)

File: apps/system/utils/utils_heapinfo.c [hunk 2]
```diff
@@ -269,7 +261,22 @@ int utils_heapinfo(int argc, char **args)
...
-	struct mm_heap_s *heap = g_mmheap;
+#ifdef CONFIG_MM_KERNEL_HEAP
+	struct mm_heap_s *heap = g_kmmheap;
+#else
+	void *temp_addr = malloc(1);
+	if (temp_addr == NULL) {
+		printf("Fail to run heapinfo : Out of memory.\n");
+		return ERROR;
+	}
+	struct mm_heap_s *heap = mm_get_heap(temp_addr);
```
Comment (junmin-kim) [inline apps/system/utils/utils_heapinfo.c]: if `CONFIG_MM_KERNEL_HEAP` is defined, `heap` will be redefined.

File: apps/system/utils/utils_heapinfo.c [hunk 3]
```diff
@@ -269,7 +261,22 @@ int utils_heapinfo(int argc, char **args)
...
+	if (heap == NULL) {
+		printf("Fail to run heapinfo : Operation fail.\n");
```
Comment (SeonghoByeon) [inline apps/system/utils/utils_heapinfo.c]: Would temp_addr be freed in mm_get_heap() when heap is null?

Comment (jeongchanKim) [reply apps/system/utils/utils_heapinfo.c]: Done. Thanks.

**Comment by seinfra** [comment]
Target : [38befc3d7f9b79facec7bc0bb1a0939f3443d60b] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [38befc3d7f9b79facec7bc0bb1a0939f3443d60b] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [a834c19b5dfbd75d6034d6984de0447e4377cca5] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [a834c19b5dfbd75d6034d6984de0447e4377cca5] - Code Rule Check OK.

File: apps/system/utils/utils_heapinfo.c [hunk 4]
```diff
@@ -269,7 +261,22 @@ int utils_heapinfo(int argc, char **args)
...
 	heapinfo_init_totalinfo();
 #endif
-	struct mm_heap_s *heap = g_mmheap;
+#ifdef CONFIG_MM_KERNEL_HEAP
+	struct mm_heap_s *heap = g_kmmheap;
+#else
+	void *temp_addr = malloc(1);
+	if (temp_addr == NULL) {
+		printf("Fail to run heapinfo : Out of memory.\n");
+		return ERROR;
```
Comment (yashwanth686007) [inline apps/system/utils/utils_heapinfo.c]: ERROR->ENOMEM?

File: apps/system/utils/utils_heapinfo.c [hunk 5]
```diff
@@ -269,7 +261,22 @@ int utils_heapinfo(int argc, char **args)
...
+	}
+	struct mm_heap_s *heap = mm_get_heap(temp_addr);
+	free(temp_addr);
+	if (heap == NULL) {
+		printf("Fail to run heapinfo : Operation fail.\n");
+		return ERROR;
```
Comment (yashwanth686007) [inline apps/system/utils/utils_heapinfo.c]: Better to return appropriate return value.

**Comment by seinfra** [comment]
Target : [1a972b0d150d847124cf6c661c6000204327d719] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1a972b0d150d847124cf6c661c6000204327d719] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [4480fa975393e2cae6ad109d4fe0366bf5feca45] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [4480fa975393e2cae6ad109d4fe0366bf5feca45] - Code Rule Check (C++) OK.

**Pull request merged**